codegen/smithy-ruby-codegen/src/main/java/software/amazon/smithy/ruby/codegen/middleware/factories/AuthMiddlewareFactory.java

```diff
@@ -0,0 +1,133 @@
+            clientConfigSet.add(ClientConfig.builder()
+                    .name(identityResolverConfigNameForAuthTrait(trait))
+                    .type(Hearth.IDENTITY_RESOLVER.toString())
+                    .documentation(identityResolverDocumentation.formatted(Hearth.IDENTITY_RESOLVER, shapeId))
+                    .defaultDynamicValue(defaultIdentityResolverForAuthTrait(trait))
+                    .allowOperationOverride()
+                    .build());
+        });
+
+        String authResolverDocumentation = """
+                A class that responds to a `resolve(auth_params)` method where `auth_params` is
+                the {Auth::Params} struct. For a given operation_name, the method must return an
+                ordered list of {%s} objects to be considered for authentication.
+                """.formatted(Hearth.AUTH_OPTION);
+        ClientConfig authResolver = ClientConfig.builder()
+                .name("auth_resolver")
+                .type("Auth::Resolver")
+                .allowOperationOverride()
+                .documentation(authResolverDocumentation)
+                .defaultValue("Auth::Resolver.new")
+                .build();
+
+        String authSchemesDocumentation = """
+                An ordered list of {%s} objects that will considered when attempting to authenticate
+                the request. The first scheme that returns an Identity from its %s will be used to
+                authenticate the request.
+                """.formatted(Hearth.AUTH_SCHEMES + "::Base", Hearth.IDENTITY_RESOLVER);
+        ClientConfig authSchemes = ClientConfig.builder()
+                .name("auth_schemes")
+                .type("Array")
+                .documentationType("Array<" + Hearth.AUTH_SCHEMES + "::Base>")
+                .defaultValue("Auth::SCHEMES")
+                .allowOperationOverride()
+                .documentation(authSchemesDocumentation)
+                .build();
+
+        Middleware.Builder authBuilder = Middleware.builder()
+                .klass(Hearth.AUTH_MIDDLEWARE)
+                .step(MiddlewareStackStep.FINALIZE)
+                .addConfig(authResolver)
+                .addConfig(authSchemes)
+                .operationParams((ctx, operation) -> {
+                    Map<String, String> params = new HashMap<>();
+                    Symbol symbol = symbolProvider.toSymbol(operation);
+                    String operationName = RubyFormatter.toSnakeCase(symbol.getName());
+                    // TODO: support more auth params in the future
+                    String authParams = "Auth::Params.new(operation_name: :%s)".formatted(operationName);
+                    params.put("auth_params", authParams);
+                    return params;
+                });
+        clientConfigSet.forEach(authBuilder::addConfig);
+
+        return authBuilder.build();
+    }
+
+    private static String identityResolverConfigNameForAuthTrait(Trait trait) {
```

Comment:
I think we could clean up a lot of these If/else with some sort of AuthScheme Codegen class(es).